### ISSUE TITLE
Bitcoind to use RPC authentication cookie from file

### ISSUE THREAD

**Issue by schildbach** [bitcoin/bitcoin#20387]
I'm running bitcoind 0.20.1 inside a Docker container and would like to access its RPC interface from another container.

Docker has a way to manage secrets like the RPC authentication cookie and safely pass them between containers. Effectively, it mounts the secret as a file like `/run/secrets/rpc_cookie`. I'd like bitcoind to use the cookie from that file.

I tried `-rpccookiefile=/run/secrets/rpc_cookie`, but bitcoind tries to write to that file and fails. Of course, it's read-only.

Alternatives: Docker doesn't support to insert secrets (which can be any data, including binary) into configuration files or to pass it as a command line argument (which would leak the cookie  anyway, I guess).

**Proposal:** I propose adding a mode in which bitcoind reads the authentication cookie from file, rather than trying to create a new one. That mode could be auto-enabled if the specified cookie file exists on startup and is read-only.

Side note: Because of at least one consumer of the RPC API (electrs) only supports reading cookies from file (both command like arguments and environment variables are deemed insecure), I will now have to define the same secret at two places – which I think has a higher risk to leak the secret.

**Comment by ryanofsky** [comment]
This seems like a good thing to support, but maybe it would be cleaner to use password authentication instead of cookie authentication here. Cookies being secrets chosen by the server, and passwords being secrets chosen externally. But if not getting caught up in cookie/password semantics, we could add a `-rpccookiegenerate`/`-norpccookiegenerate` boolean option defaulting to true, to toggle between either generating the cookie internally or allowing a cookie value to be specified externally. I think adding an option to use external cookies would be better than automatically trusting read-only cookie files, which might be created unintentionally or maliciously.

Another suggestion to solve the immediate problem of using bitcoin and electrs with docker secrets: maybe you could have docker generate `/run/secrets/bitcoin_password` and have a startup script include it in electrs and bitcoind configurations (bitcoin `-includeconf` option might be useful here).

**Comment by promag** [comment]
```
  -conf=<file>
       Specify path to read-only configuration file. Relative paths will be
       prefixed by datadir location. (default: bitcoin.conf)
```
Make the whole conf file a secret with desired `rpcauth=...` credentials. Same for electrs.

**Comment by schildbach** [comment]
@ryanofsky I thought rpc passwords are going to go away? They're already deprecated in current bitcoind. I agree with your concern about the implicit toggle.

@promag I thought of that too, but putting the entire configuration into a secret seems like an abuse of the secrets mechanism. For example, conf files do not support Docker compose variable substitution, as opposed to config via arguments or environment variables.

**Comment by sipa** [comment]
@schildbach `-rpcpassword`/`-rpcuser` is deprecated. The new mechanism is `-rpcauth`, which uses the same internal mechanism as the cookie logic (but supports configuring using externally generated passwords).

**Comment by promag** [comment]
> conf files do not support Docker compose variable substitution, as opposed to config via arguments or environment variables.

@schildbach those can be set as the command (beside `-conf`) right?

**Comment by schildbach** [comment]
> @schildbach `-rpcpassword`/`-rpcuser` is deprecated. The new mechanism is `-rpcauth`, which uses the same internal mechanism as the cookie logic (but supports configuring using externally generated passwords).

Exactly, and this feature request is about reading the RPC auth cookie from a file.

**Comment by promag** [comment]
@schildbach hare you tried to use docker compose `volumes_from`?

**Comment by promag** [comment]
Anyway, I'm concept NACK on this. I think we could support `-rpcauth=path` instead.

**Comment by MarcoFalke** [comment]
> but putting the entire configuration into a secret seems like an abuse of the secrets mechanism

It doesn't have to be the whole config. It can be just an `-includeconf` which contains just one line with the rpcauth blob

**Comment by schildbach** [comment]
Thanks for the hint. It still feels kludgy and unsafe though. I don't want the secret to be able to add arbitrary configuration to my bitcoind. Note that secrets are not meant to be reviewed, while configuration is like code that should be seen by as many eyes as possible.

**Comment by promag** [comment]
@schildbach feel free to test #20407.

**Comment by alevchuk** [comment]
This seems to go against the [purpose of the cookie file](https://github.com/Kixunil/security_writings/blob/master/cookie_files.md#the-problem-with-passwords). To avoid another account capturing the password, the cookie file needs to be regenerated every time bitcoind restarts. Also, it needs to be generated after `bitcoind` binds to the RPC port, so the best place to do generation is inside `bitcoind`

**Comment by thinkmassive** [comment]
> putting the entire configuration into a secret seems like an abuse of the secrets mechanism.

It seems like `rpcauthfile` (proposed in #20407) could be valuable in a Kubernetes environment, where the cookie approach might be less viable, and storing the `rpcauth` string in plaintext in the config file is obviously less secure (but likely, when the entire config is stored as a secret).

The `rpcauth` string could be stored as a secret on its own, which can be fetched and provided for one-time use at startup [from a named pipe](https://github.com/lightninglabs/lndinit#5-store-the-wallet-password-in-a-file) that disappears from the filesystem after being read once. This approach seems even more secure than using a cookie file, which persists on the filesystem during operation.

**Comment by willcl-ark** [comment]
It seems like the consensus is that one should use `rpcauth` in this situation.

Let us know if you feel this should be reopened.
